### DISCUSSION TITLE
Update Migration Guide for `add_foreign_key`

### DISCUSSION THREAD

**Discussion by jvanderen1** [jeremyevans/sequel#2202] [General]
When using `add_foreign_key` in a migration within a `change` block and an array with 1 symbol, this does appear to go `down` properly. Therefore, updating the [migration guide](https://sequel.jeremyevans.net/rdoc/files/doc/migration_rdoc.html#:~:text=add_constraint-,add_foreign_key,-\(with%20a%20symbol) might be helpful for others.

Example:

```ruby
Sequel.migration do
  change do
    alter_table :projects do
      add_foreign_key([:user_id], :users, not_valid: true)
    end
  end
end
```

- Migrating this `up`: Within the the `projects`, this creates a `user_id` column + foreign key constraint pointing to the `users` table (invalid)
- Migrating this `down`: Within the the `projects`, this drops the foreign key constraint within the `user_id` column pointing to the `users` table

**Comment by jeremyevans** [comment]
When you pass an array of columns, Sequel should only attempt to create the foreign key constraint, not add a column:

```ruby
DB.alter_table :projects do
  add_foreign_key([:user_id], :users, not_valid: true)
end
# ALTER TABLE "projects" ADD FOREIGN KEY ("user_id") REFERENCES "users" NOT VALID
```

This isn't reversible as you did not name the foreign key constraint.  Can you show an example of Sequel creating a column when passing an array of columns?

**Reply by jvanderen1** [reply]
Absolutely! Here is a minimal example which:
- Has a `user_id` column which already exists
- Has no foreign key constraint

## Example

Given a `projects` table schema which looks like:

```
                                               Table "public.projects"
           Column            |            Type             | Collation | Nullable |             Default
-----------------------------+-----------------------------+-----------+----------+----------------------------------
 id                          | integer                     |           | not null | generated by default as identity
 user_id                     | integer                     |           | not null |
Indexes:
    "projects_pkey" PRIMARY KEY, btree (id)
```

I run the example migration in the [description](https://github.com/jeremyevans/sequel/discussions/2202#discussion-7046669) `up`. This generates a foreign key constraint:

```
                                               Table "public.projects"
           Column            |            Type             | Collation | Nullable |             Default
-----------------------------+-----------------------------+-----------+----------+----------------------------------
 id                          | integer                     |           | not null | generated by default as identity
 user_id                     | integer                     |           | not null |
Indexes:
    "projects_pkey" PRIMARY KEY, btree (id)
Foreign key constraints:
    "projects_user_id_fkey"  | (user_id) REFERENCES users(id) NOT VALID
```

I run the example migration in the [description](https://github.com/jeremyevans/sequel/discussions/2202#discussion-7046669) `down`. This removes the foreign key constraint:

```
                                               Table "public.projects"
           Column            |            Type             | Collation | Nullable |             Default
-----------------------------+-----------------------------+-----------+----------+----------------------------------
 id                          | integer                     |           | not null | generated by default as identity
 user_id                     | integer                     |           | not null |
Indexes:
    "projects_pkey" PRIMARY KEY, btree (id)
```

## Note

To your point:

> Sequel should only attempt to create the foreign key constraint, not add a column

If the column `user_id` did not already exist, the migration would actually fail at `up`! So there are specific cases where an array won't work.

**Reply by jeremyevans** [reply]
From my testing, it looks like it works in the cases where:

* The database supports `foreign_key_list`
* There is only a single foreign key constraint for the columns

In this case, Sequel can look up the foreign key name and drop the constraint. This appears to be supported on common databases, but not on all databases that Sequel ships support for.  I would guess multiple foreign key constraints on the same columns are rare.  So for most Sequel usage, it probably does work.

I'll update the documentation to reflect that.  Thank you very much for your perseverance.

**Reply by jvanderen1** [reply]
Thanks @jeremyevans!
